coderd/tailnet.go

```diff
@@ -0,0 +1,340 @@
+package coderd
+
+import (
+	"bufio"
+	"context"
+	"net"
+	"net/http"
+	"net/http/httputil"
+	"net/netip"
+	"net/url"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/google/uuid"
+	"golang.org/x/xerrors"
+	"tailscale.com/derp"
+	"tailscale.com/tailcfg"
+
+	"cdr.dev/slog"
+	"github.com/coder/coder/coderd/wsconncache"
+	"github.com/coder/coder/codersdk"
+	"github.com/coder/coder/site"
+	"github.com/coder/coder/tailnet"
+)
+
+var defaultTransport *http.Transport
+
+func init() {
+	var valid bool
+	defaultTransport, valid = http.DefaultTransport.(*http.Transport)
+	if !valid {
+		panic("dev error: default transport is the wrong type")
+	}
+}
+
+// TODO: ServerTailnet does not currently remove stale peers.
+
+// NewServerTailnet creates a new tailnet intended for use by coderd. It
+// automatically falls back to wsconncache if a legacy agent is encountered.
+func NewServerTailnet(
+	ctx context.Context,
+	logger slog.Logger,
+	derpServer *derp.Server,
+	derpMap *tailcfg.DERPMap,
+	coord *atomic.Pointer[tailnet.Coordinator],
+	cache *wsconncache.Cache,
+) (*ServerTailnet, error) {
+	conn, err := tailnet.NewConn(&tailnet.Options{
+		Addresses: []netip.Prefix{netip.PrefixFrom(tailnet.IP(), 128)},
+		DERPMap:   derpMap,
+		Logger:    logger.Named("tailnet"),
+	})
+	if err != nil {
+		return nil, xerrors.Errorf("create tailnet conn: %w", err)
+	}
+
+	tn := &ServerTailnet{
+		logger:      logger,
+		conn:        conn,
+		coordinator: coord,
+		cache:       cache,
+		agentNodes:  map[uuid.UUID]*tailnetNode{},
+		transport:   defaultTransport.Clone(),
+	}
+	tn.transport.DialContext = tn.dialContext
+	tn.transport.MaxIdleConnsPerHost = 10
+	tn.transport.MaxIdleConns = 0
+
+	conn.SetNodeCallback(func(node *tailnet.Node) {
+		tn.nodesMu.Lock()
+		ids := make([]uuid.UUID, 0, len(tn.agentNodes))
+		for id := range tn.agentNodes {
+			ids = append(ids, id)
+		}
+		tn.nodesMu.Unlock()
+
+		err := (*tn.coordinator.Load()).BroadcastToAgents(ids, node)
+		if err != nil {
+			tn.logger.Error(context.Background(), "broadcast server node to agents", slog.Error(err))
+		}
+	})
+
+	// This is set to allow local DERP traffic to be proxied through memory
+	// instead of needing to hit the external access URL. Don't use the ctx
+	// given in this callback, it's only valid while connecting.
+	conn.SetDERPRegionDialer(func(_ context.Context, region *tailcfg.DERPRegion) net.Conn {
+		if !region.EmbeddedRelay {
+			return nil
+		}
+		left, right := net.Pipe()
+		go func() {
+			defer left.Close()
+			defer right.Close()
+			brw := bufio.NewReadWriter(bufio.NewReader(right), bufio.NewWriter(right))
+			derpServer.Accept(ctx, right, brw, "internal")
+		}()
+		return left
+	})
+
+	return tn, nil
+}
+
+type tailnetNode struct {
+	node           *tailnet.Node
+	lastConnection time.Time
+	stop           func()
+}
+
+type ServerTailnet struct {
```

Comment:
Should we toss this in the `tailnet` package for now? Or can we not because of the `wsconncache` dep?

Reply:
Yeah, putting this in `tailnet` causes cyclical imports

Reply:
Reasonable